manager.go

```diff
@@ -0,0 +1,248 @@
+// membership of _name.
+//
+// The name itself must not start with an underscore.
+//
+// This does nothing if an ACL with that name already exists.
+func (h *Manager) CreateACL(ctx context.Context, name string, initialUsers ...string) error {
+	if isMetaName(name) {
+		return errgo.Newf("invalid ACL name %q", name)
+	}
+	if err := h.p.Store.CreateACL(ctx, name, initialUsers...); err != nil {
+		return errgo.Mask(err, errgo.Is(ErrBadUsername))
+	}
+	if err := h.p.Store.CreateACL(ctx, metaName(name)); err != nil {
+		return errgo.Mask(err)
+	}
+	return nil
+}
+
+// aclName is implemented by the request parameters for all endpoints
+// to return the associated ACL name.
+type aclName interface {
+	ACLName() string
+}
+
+type handler struct {
+	manager *Manager
+}
+
+// newHandler returns a handler instance to serve a particular HTTP request.
+func (m *Manager) newHandler(p httprequest.Params, arg aclName) (handler, context.Context, error) {
+	ctx := p.Context
+	if err := m.authorizeRequest(ctx, p, arg.ACLName()); err != nil {
+		return handler{}, nil, errgo.Mask(err, errgo.Any)
+	}
+	return handler{
+		manager: m,
+	}, p.Context, nil
+}
+
+// authorizeRequest checks that an HTTP request is authorized. If the
+// authorization failed because Authenticate failed, it returns an error
+// with an errAuthenticationFailed cause to signal that the desired
+// error response has already been written.
+func (m *Manager) authorizeRequest(ctx context.Context, p httprequest.Params, aclName string) error {
+	if aclName == "" {
+		return httprequest.Errorf(httprequest.CodeBadRequest, "empty ACL name")
+	}
+	identity, err := m.p.Authenticate(ctx, p.Response, p.Request)
+	if err != nil {
+		return errAuthenticationFailed
+	}
+	var checkACLName string
+	if aclName == AdminACL || isMetaName(aclName) {
+		// We're trying to access either the admin ACL or a meta-ACL; for either
+		// of these, admin privileges are needed.
+		checkACLName = AdminACL
+	} else {
+		// For all normal ACLs, access for a given ACL name is decided via membership
+		// of the meta-ACL for that name.
+		checkACLName = metaName(aclName)
+	}
+	acl, err := m.ACL(ctx, checkACLName)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(ErrACLNotFound))
+	}
+	if checkACLName != AdminACL {
+		// Admin users always get permission to do anything.
+		adminACL, err := m.ACL(ctx, AdminACL)
+		if err != nil {
+			return errgo.Notef(err, "cannot get admin ACL")
+		}
+		acl = append(acl, adminACL...)
+	}
+	ok, err := identity.Allow(ctx, acl)
+	if err != nil {
+		return errgo.Notef(err, "cannot check permissions")
+	}
+	if !ok {
+		return httprequest.Errorf(httprequest.CodeForbidden, "")
+	}
+	return nil
+}
+
+// Get returns the members of the ACL with the requested name.
+// Only administrators and members of the meta-ACL for the name
+// may access this endpoint. The meta-ACL for meta-ACLs is "admin".
+func (h handler) Get(p httprequest.Params, req *params.GetRequest) (*params.GetResponse, error) {
```

Comment:
Get is a terrible name it's too overloaded.

Reply:
changed